src/Streamly/Internal/Data/Unfold/Enumeration.hs

```diff
@@ -0,0 +1,336 @@
+-- Enumeration of Fractionals
+------------------------------------------------------------------------------
+
+-- | /Internal/
+--
+-- > enumerateFromToFractional to = takeWhile (<= to + 1 / 2) $ enumerateFromStepNum 1
+--
+{-# INLINE_NORMAL enumerateFromToFractional #-}
+enumerateFromToFractional :: (Monad m, Fractional a, Ord a) => a -> Unfold m a a
+enumerateFromToFractional to =
+    takeWhile (<= to + 1 / 2) $ enumerateFromStepNum 1
+
+{-# INLINE_NORMAL enumerateFromFractional #-}
+enumerateFromFractional :: (Monad m, Fractional a) => Unfold m a a
+enumerateFromFractional =  enumerateFromStepNum 1
+
+{-# INLINE_NORMAL enumerateFromThenToFractional #-}
+enumerateFromThenToFractional :: (Monad m, Fractional a, Ord a) => a -> a -> Unfold m a a
+enumerateFromThenToFractional stride to =
+    takeWhile (<= to + 1 / 2) $ enumerateFromStepNum stride
+
+{-# INLINE_NORMAL enumerateFromThenFractional #-}
+enumerateFromThenFractional :: (Monad m, Fractional a) => a -> Unfold m a a
+enumerateFromThenFractional = enumerateFromStepNum
+
+-------------------------------------------------------------------------------
+-- Enumerable type class
+-------------------------------------------------------------------------------
+-- | Types that can be enumerated as a stream. The operations in this type
+-- class are equivalent to those in the 'Enum' type class, except that these
+-- generate a stream instead of a list. Use the functions in
+-- "Streamly.Internal.Data.Unfold.Enumeration" module to define new instances.
+--
+-- /Pre-release/
+class Enum a => Enumerable a where
+
+-- @enumerateFrom from@ generates a stream starting with the element
+-- @from@, enumerating up to 'maxBound' when the type is 'Bounded' or
+-- generating an infinite stream when the type is not 'Bounded'.
+--
+-- @
+-- >>> Stream.toList $ Stream.take 4 $ Stream.unfold Unfold.enumerateFrom (0 :: Int)
+-- [0,1,2,3]
+--
+-- @
+--
+-- For 'Fractional' types, enumeration is numerically stable. However, no
+-- overflow or underflow checks are performed.
+--
+-- @
+-- >>> Stream.toList $ Stream.take 4 $ Stream.unfold Unfold.enumerateFrom 1.1
+-- [1.1,2.1,3.1,4.1]
+--
+-- @
+--
+-- /Pre-release/
+--
+    enumerateFrom :: (Integral a, Monad m) => Unfold m a a
+
+-- Generate a finite stream starting with the element @from@, enumerating
+-- the type up to the value @to@. If @to@ is smaller than @from@ then an
+-- empty stream is returned.
+--
+-- @
+-- >>> Stream.toList $ Stream.unfold (Unfold.enumerateFromTo 4) 0
+-- [0,1,2,3,4]
+--
+-- @
+--
+-- For 'Fractional' types, the last element is equal to the specified @to@
+-- value after rounding to the nearest integral value.
+--
+-- @
+-- >>> Stream.toList $ Stream.unfold (Unfold.enumerateFromTo 4) 1.1
+-- [1.1,2.1,3.1,4.1]
+--
+-- >>> Stream.toList $ Stream.unfold (Unfold.enumerateFromTo 4.6) 1.1
+-- [1.1,2.1,3.1,4.1,5.1]
+--
+-- @
+--
+-- /Pre-release/
+    enumerateFromTo :: (Integral a, Monad m) => a -> Unfold m a a
```

Comment:
I think, the signature should be `Unfold (a, a) a`. We should move all the inputs to unfold.

Reply:
We should change the signatures of `enumerateFromThenTo` etc functions defined earlier as well in the same way.